File: LeetCode/201-300/221. Maximal Square/src/Solution.java
class Solution {
    public int maximalSquare(char[][] matrix) {
        int ans = 0;
        for (int j = 0; j < matrix[0].length; j++) {
            matrix[0][j] -= '0';
            if (ans == 0 && matrix[0][j] == 1) {
                ans = 1;
            }
        }
        for (int i = 1; i < matrix.length; i++) {
            matrix[i][0] -= '0';
            if (ans == 0 && matrix[i][0] == 1) {
                ans = 1;
            }
            for (int j = 1; j < matrix[0].length; j++) {
                if (matrix[i][j] == '1') {
                    matrix[i][j] = (char) (Math.min(matrix[i - 1][j - 1], Math.min(matrix[i][j - 1], matrix[i - 1][j])) + 1);
                    ans = Math.max(ans, matrix[i][j]);
                } else {
                    matrix[i][j] = 0;
                }
            }
        }
        return ans * ans;
    }
}

File: LeetCode/201-300/221. Maximal Square/Solution.cpp
class Solution {
public:
    int maximalSquare(vector<vector<char>>& matrix) {
        int m = matrix.size();
        int n = matrix[0].size();
        int maxSq = 0;
        int dp[n];
        int leftUp;
        int temp;
        for (int j = 0; j < n; ++j) {
            dp[j] = matrix[0][j] - '0';
            if (dp[j] > 0) {
                maxSq = 1;
            }
        }
        for (int i = 1; i < m; ++i) {
            temp = matrix[i][0] - '0';
            maxSq = max(maxSq, temp);
            leftUp = dp[0];
            dp[0] = temp;
            for (int j = 1; j < n; ++j) {
                temp = min(dp[j - 1], min(leftUp, dp[j]));
                leftUp = dp[j];
                dp[j] = (matrix[i][j] == '0') ? 0 : temp + 1;
                maxSq = max(maxSq, dp[j]);
            }
        }
        return maxSq * maxSq;
    }
};